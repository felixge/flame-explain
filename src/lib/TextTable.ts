import {Node as FNode} from './FlameExplain';
import {formatDuration} from './Util';
// @ts-ignore no type definitions
import AsciiTable from 'ascii-table';

export type Column =
  '#' |
  'Actual Loops' |
  'Actual Rows' |
  'Actual Total Time' |
  'Label' |
  'Node Type' |
  'Self Time' |
  'Total Time' |
  'Virtual';

export function textTable(
  n: FNode,
  {columns = [], title = ''}: {
    columns: Column[],
    title?: string
  }
): string {
  let table = new AsciiTable(title);
  table.setHeading(...columns);

  columns.forEach((col, i) => {
    if (!(col === '#' || col === 'Label')) {
      table.setAlignRight(i);
    }
  });

  let rowNum = 1;
  const visit = (n: FNode, depth = 0) => {
    if (!n.Root) {
      const colVals = columns.map(c => extractColumn(n, c, rowNum, depth));
      table.addRow(...colVals);
      rowNum++;
      depth++;
    }

    (n.Children || []).forEach(child => visit(child, depth));
  }
  visit(n);

  return table.toString();
}

export function extractColumn(n: FNode, c: Column, rowNum: number, depth: number): any {
  switch (c) {
    case '#':
      return rowNum;
    case 'Actual Loops':
      return ('Actual Loops' in n.Source)
        ? n.Source["Actual Loops"]
        : '';
    case 'Actual Rows':
      return ('Actual Rows' in n.Source)
        ? n.Source["Actual Rows"]
        : '';
    case 'Actual Total Time':
      return ('Actual Total Time' in n.Source)
        ? formatDuration(n.Source["Actual Total Time"])
        : '';
    case 'Label':
      return '  '.repeat(depth) + n.Label;
    case 'Node Type':
      return ('Node Type' in n.Source)
        ? n.Source["Node Type"]
        : '';
    case 'Self Time':
      return ('Self Time' in n)
        ? formatDuration(n["Self Time"])
        : '';
    case 'Total Time':
      return ('Total Time' in n)
        ? formatDuration(n["Total Time"])
        : '';
    case 'Virtual':
      return (n.Virtual)
        ? 'x'
        : '';
  }
}
