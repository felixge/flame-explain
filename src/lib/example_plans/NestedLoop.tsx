import { ExamplePlan } from "./"

const Sample: ExamplePlan = {
  sql: `
/** A trivial nested loop plan. FlameExplain correctly accounts for the
"Actual Total Time" in the looped node to be the average time per loop. It
also highlights the high cost of query planning relative to the execution of
this plan. */
EXPLAIN (ANALYZE, FORMAT JSON)
SELECT *
FROM generate_series(1, 10) a
JOIN generate_series(5, 15) b ON a > b;
`,
  queries: [
    {
      "Plan": {
        "Node Type": "Nested Loop",
        "Parallel Aware": false,
        "Join Type": "Inner",
        "Startup Cost": 0.01,
        "Total Cost": 2.58,
        "Plan Rows": 37,
        "Plan Width": 8,
        "Actual Startup Time": 0.037,
        "Actual Total Time": 0.059,
        "Actual Rows": 15,
        "Actual Loops": 1,
        "Inner Unique": false,
        "Join Filter": "(a.a > b.b)",
        "Rows Removed by Join Filter": 95,
        "Plans": [
          {
            "Node Type": "Function Scan",
            "Parent Relationship": "Outer",
            "Parallel Aware": false,
            "Function Name": "generate_series",
            "Alias": "a",
            "Startup Cost": 0.0,
            "Total Cost": 0.1,
            "Plan Rows": 10,
            "Plan Width": 4,
            "Actual Startup Time": 0.008,
            "Actual Total Time": 0.009,
            "Actual Rows": 10,
            "Actual Loops": 1,
          },
          {
            "Node Type": "Function Scan",
            "Parent Relationship": "Inner",
            "Parallel Aware": false,
            "Function Name": "generate_series",
            "Alias": "b",
            "Startup Cost": 0.0,
            "Total Cost": 0.11,
            "Plan Rows": 11,
            "Plan Width": 4,
            "Actual Startup Time": 0.001,
            "Actual Total Time": 0.002,
            "Actual Rows": 11,
            "Actual Loops": 10,
          },
        ],
      },
      "Planning Time": 0.067,
      "Triggers": [],
      "Execution Time": 0.087,
    },
  ],
}
export default Sample
