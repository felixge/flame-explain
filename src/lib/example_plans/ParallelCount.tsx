import { ExamplePlan } from './';

const Sample: ExamplePlan = {
  sql: `/** A simple count(*) query that uses a parallel sequential scan. */
CREATE TABLE foo AS
SELECT g
FROM generate_series(1, 10000000) g;

EXPLAIN (ANALYZE, FORMAT JSON)
SELECT count(*)
FROM foo;
`,
  queries: [
    {
      'Plan': {
        'Node Type': 'Aggregate',
        'Strategy': 'Plain',
        'Partial Mode': 'Finalize',
        'Parallel Aware': false,
        'Startup Cost': 97331.43,
        'Total Cost': 97331.44,
        'Plan Rows': 1,
        'Plan Width': 8,
        'Actual Startup Time': 834.738,
        'Actual Total Time': 834.738,
        'Actual Rows': 1,
        'Actual Loops': 1,
        'Plans': [
          {
            'Node Type': 'Gather',
            'Parent Relationship': 'Outer',
            'Parallel Aware': false,
            'Startup Cost': 97331.21,
            'Total Cost': 97331.42,
            'Plan Rows': 2,
            'Plan Width': 8,
            'Actual Startup Time': 834.619,
            'Actual Total Time': 850.904,
            'Actual Rows': 3,
            'Actual Loops': 1,
            'Workers Planned': 2,
            'Workers Launched': 2,
            'Single Copy': false,
            'Plans': [
              {
                'Node Type': 'Aggregate',
                'Strategy': 'Plain',
                'Partial Mode': 'Partial',
                'Parent Relationship': 'Outer',
                'Parallel Aware': false,
                'Startup Cost': 96331.21,
                'Total Cost': 96331.22,
                'Plan Rows': 1,
                'Plan Width': 8,
                'Actual Startup Time': 830.591,
                'Actual Total Time': 830.591,
                'Actual Rows': 1,
                'Actual Loops': 3,
                'Plans': [
                  {
                    'Node Type': 'Seq Scan',
                    'Parent Relationship': 'Outer',
                    'Parallel Aware': true,
                    'Relation Name': 'foo',
                    'Alias': 'foo',
                    'Startup Cost': 0.0,
                    'Total Cost': 85914.57,
                    'Plan Rows': 4166657,
                    'Plan Width': 0,
                    'Actual Startup Time': 0.019,
                    'Actual Total Time': 473.428,
                    'Actual Rows': 3333333,
                    'Actual Loops': 3,
                  },
                ],
              },
            ],
          },
        ],
      },
      'Planning Time': 0.054,
      'Triggers': [],
      'Execution Time': 850.961,
    },
  ],
};
export default Sample;
