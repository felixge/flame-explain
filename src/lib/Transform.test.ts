import {textNodeName, transformQueries} from './Transform'
import {textTable, Column, queryFirst} from './TextTable';
import examplePlans from './example_plans';
import NestedLoop from './example_plans/NestedLoop';
import ParallelCount from './example_plans/ParallelCount';

describe('textNodeName', () => {
  test('Result', () => {
    expect(textNodeName({"Node Type": 'Result'})).toEqual('Result');
  });

  test('ModifyTable', () => {
    expect(textNodeName({
      "Node Type": 'ModifyTable',
      "Operation": "Insert",
      'Relation Name': '',
      'Alias': '',
    })).toEqual('Insert');
    expect(textNodeName({
      "Node Type": 'ModifyTable',
      "Operation": "Update",
      'Relation Name': '',
      'Alias': '',
    })).toEqual('Update');
    expect(textNodeName({
      "Node Type": 'ModifyTable',
      "Operation": "Delete",
      'Relation Name': '',
      'Alias': '',
    })).toEqual('Delete');

    expect(textNodeName({
      "Node Type": 'ModifyTable',
      "Operation": "Insert",
      'Relation Name': 'foo',
      'Alias': '',
    })).toEqual('Insert on foo');
    expect(textNodeName({
      "Node Type": 'ModifyTable',
      "Operation": "Insert",
      'Relation Name': 'foo',
      'Alias': 'bar',
    })).toEqual('Insert on foo bar');
    expect(textNodeName({
      "Node Type": 'ModifyTable',
      "Operation": "Insert",
      'Relation Name': 'foo',
      'Alias': 'bar',
      'Schema': 'baz',
    })).toEqual('Insert on baz.foo bar');
  });

  test('Foreign Scan', () => {
    expect(textNodeName({
      "Node Type": 'Foreign Scan',
      "Operation": "Select",
      'Relation Name': '',
      'Alias': '',
    })).toEqual('Foreign Scan');
    expect(textNodeName({
      "Node Type": 'Foreign Scan',
      "Operation": "Insert",
      'Relation Name': '',
      'Alias': '',
    })).toEqual('Foreign Insert');
    expect(textNodeName({
      "Node Type": 'Foreign Scan',
      "Operation": "Update",
      'Relation Name': '',
      'Alias': '',
    })).toEqual('Foreign Update');
    expect(textNodeName({
      "Node Type": 'Foreign Scan',
      "Operation": "Delete",
      'Relation Name': '',
      'Alias': '',
    })).toEqual('Foreign Delete');
  });

  test('Join Type', () => {
    expect(textNodeName({
      "Node Type": 'Hash Join',
      "Hash Cond": '',
      "Join Type": "Left",
    })).toEqual('Hash Left Join');
    expect(textNodeName({
      "Node Type": 'Merge Join',
      "Join Type": "Left",
    })).toEqual('Merge Left Join');
    expect(textNodeName({
      "Node Type": 'Hash Join',
      "Hash Cond": '',
      "Join Type": "Inner",
    })).toEqual('Hash Join');
    expect(textNodeName({
      "Node Type": 'Nested Loop',
      "Join Type": "Inner",
    })).toEqual('Nested Loop');
    expect(textNodeName({
      "Node Type": 'Nested Loop',
      "Join Type": "Left",
    })).toEqual('Nested Loop Left Join');
  });

  test('Aggregate', () => {
    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "Plain",
    })).toEqual('Aggregate');
    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "Sorted",
    })).toEqual('GroupAggregate');
    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "Hashed",
    })).toEqual('HashAggregate');
    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "Mixed",
    })).toEqual('MixedAggregate');
    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "???",
    })).toEqual('Aggregate ???');

    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "Sorted",
      "Partial Mode": "Simple",
    })).toEqual('GroupAggregate');
    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "Sorted",
      "Partial Mode": "Partial",
    })).toEqual('Partial GroupAggregate');
    expect(textNodeName({
      "Node Type": 'Aggregate',
      "Strategy": "Sorted",
      "Partial Mode": "Finalize",
    })).toEqual('Finalize GroupAggregate');
  });

  test('SetOp', () => {
    expect(textNodeName({
      "Node Type": 'SetOp',
      "Strategy": "Sorted",
      "Command": "Intersect",
    })).toEqual('SetOp Intersect');
    expect(textNodeName({
      "Node Type": 'SetOp',
      "Strategy": "Hashed",
      "Command": "Intersect All",
    })).toEqual('HashedSetOp Intersect All');
    expect(textNodeName({
      "Node Type": 'SetOp',
      "Strategy": "???",
      "Command": "Except",
    })).toEqual('SetOp ??? Except');
  });

  test('Parallel Aware', () => {
    expect(textNodeName({
      "Node Type": 'Seq Scan',
      "Parallel Aware": true,
      "Relation Name": '',
      "Alias": '',
    })).toEqual('Parallel Seq Scan');
  });

  test('PlanNodeTargetRel', () => {
    expect(textNodeName({
      "Node Type": 'Seq Scan',
      "Relation Name": 'foo',
      "Alias": '',
    })).toEqual('Seq Scan on foo');
    expect(textNodeName({
      "Node Type": 'Seq Scan',
      "Relation Name": 'foo',
      "Schema": 'myschema',
      "Alias": '',
    })).toEqual('Seq Scan on myschema.foo');
    expect(textNodeName({
      "Node Type": 'Seq Scan',
      "Relation Name": 'foo',
      "Alias": 'bar',
      "Schema": 'myschema',
    })).toEqual('Seq Scan on myschema.foo bar');

    expect(textNodeName({
      "Node Type": 'Function Scan',
      "Function Name": 'foo',
      "Schema": 'myschema',
      "Alias": 'bar',
    })).toEqual('Function Scan on myschema.foo bar');

    expect(textNodeName({
      "Node Type": 'Table Function Scan',
      "Table Function Name": 'xmltable',
      "Schema": 'myschema',
      "Alias": 'bar',
    })).toEqual('Table Function Scan on myschema.xmltable bar');

    expect(textNodeName({
      "Node Type": 'CTE Scan',
      "CTE Name": 'foo',
      "Schema": 'myschema',
      "Alias": 'bar',
    })).toEqual('CTE Scan on myschema.foo bar');
    expect(textNodeName({
      "Node Type": 'WorkTable Scan',
      "CTE Name": 'foo',
      "Schema": 'myschema',
      "Alias": 'bar',
    })).toEqual('WorkTable Scan on myschema.foo bar');

    expect(textNodeName({
      "Node Type": 'Named Tuplestore Scan',
      "Tuplestore Name": 'foo',
      "Schema": 'myschema',
      "Alias": 'bar',
    })).toEqual('Named Tuplestore Scan on myschema.foo bar');

    expect(textNodeName({
      "Node Type": 'Bitmap Index Scan',
      "Index Name": 'my_idx',
      "Index Cond": '',
    })).toEqual('Bitmap Index Scan on my_idx');

    expect(textNodeName({
      "Node Type": 'Index Scan',
      "Index Name": 'my_idx',
      "Index Cond": '',
      "Scan Direction": 'Forward',
      "Rows Removed by Index Recheck": 0,
      "Relation Name": 'foo',
      "Alias": 'bar',
      "Schema": 'myschema',
    })).toEqual('Index Scan using my_idx on myschema.foo bar');
    expect(textNodeName({
      "Node Type": 'Index Scan',
      "Index Name": 'my_idx',
      "Index Cond": '',
      "Scan Direction": 'Backward',
      "Rows Removed by Index Recheck": 0,
      "Relation Name": 'foo',
      "Alias": 'bar',
      "Schema": 'myschema',
    })).toEqual('Index Scan Backward using my_idx on myschema.foo bar');
  });
});

describe('transformQueries', () => {
  describe('time accounting snapshots are matching', () => {
    for (let name in examplePlans) {
      test(name, () => {
        const root = transformQueries(examplePlans[name].queries);
        const columns: Column[] = [
          '#',
          'Label',
          'Actual Total Time',
          'Actual Loops',
          'Total Time',
          'Self Time',
          'Virtual',
        ];
        const table = textTable(root, {title: name, columns: columns});
        expect(table).toMatchSnapshot();
      });
    }
  });

  describe('account for "Actual Loops"', () => {
    test('enabled', () => {
      const root = transformQueries(NestedLoop.queries);
      const nl = queryFirst(root, '**', 'Nested Loop');
      expect(nl).toMatchObject({
        'Actual Total Time': '59 μs',
        'Self Time': '30 μs',
        'Total Time': '59 μs',
      });
      const gsB = queryFirst(nl.Source, 'Function Scan on generate_series b');
      expect(gsB).toMatchObject({
        'Actual Loops': '10',
        'Actual Total Time': '2 μs',
        'Self Time': '20 μs',
        'Total Time': '20 μs',
      });
    });

    test('disabled', () => {
      const root = transformQueries(NestedLoop.queries, {Loops: false});
      const nl = queryFirst(root, '**', 'Nested Loop');
      expect(nl).toMatchObject({
        'Actual Total Time': '59 μs',
        'Self Time': '48 μs',
        'Total Time': '59 μs',
      });
      const gsB = queryFirst(nl.Source, 'Function Scan on generate_series b');
      expect(gsB).toMatchObject({
        'Actual Loops': '10',
        'Actual Total Time': '2 μs',
        'Self Time': '2 μs',
        'Total Time': '2 μs',
      });
    });
  });

  describe('account for parallel "Actual Loops"', () => {
    //test('enabled', () => {
    //const root = transformQueries(ParallelCount.queries, {Loops: true});
    //const fa = queryFirst(root, '**', 'Finalize Aggregate');
    //expect(fa).toMatchObject({
    //'Actual Total Time': '834.7 ms',
    //'Self Time': '0 μs',
    //'Total Time': '850.9 ms',
    //});
    //const ga = queryFirst(fa.Source, 'Gather');
    //expect(ga).toMatchObject({
    //'Actual Total Time': '850.9 ms',
    //'Self Time': '20.3 ms',
    //'Total Time': '850.9 ms',
    //});
    //const pa = queryFirst(ga.Source, 'Partial Aggregate');
    //expect(pa).toMatchObject({
    //'Actual Total Time': '830.6 ms',
    //'Self Time': '357.2 ms',
    //'Total Time': '830.6 ms',
    //});
    //const ps = queryFirst(pa.Source, 'Parallel Seq Scan on foo');
    //expect(ps).toMatchObject({
    //'Actual Total Time': '473.4 ms',
    //'Self Time': '473.4 ms',
    //'Total Time': '473.4 ms',
    //});
    //});

    //test('disabled', () => {
    //const root = transformQueries(ParallelCount.queries, {Loops: false});
    //const fa = queryFirst(root, '**', 'Finalize Aggregate');
    //expect(fa).toMatchObject({
    //'Actual Total Time': '834.7 ms',
    //'Self Time': '-16.2 ms',
    //'Total Time': '834.7 ms',
    //});
    //const ga = queryFirst(fa.Source, 'Gather');
    //expect(ga).toMatchObject({
    //'Actual Total Time': '850.9 ms',
    //'Self Time': '20.3 ms',
    //'Total Time': '850.9 ms',
    //});
    //const pa = queryFirst(ga.Source, 'Partial Aggregate');
    //expect(pa).toMatchObject({
    //'Actual Total Time': '830.6 ms',
    //'Self Time': '357.2 ms',
    //'Total Time': '830.6 ms',
    //});
    //const ps = queryFirst(pa.Source, 'Parallel Seq Scan on foo');
    //expect(ps).toMatchObject({
    //'Actual Total Time': '473.4 ms',
    //'Self Time': '473.4 ms',
    //'Total Time': '473.4 ms',
    //});
    //});
  });
});
