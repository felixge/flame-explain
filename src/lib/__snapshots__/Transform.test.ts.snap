// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformQueries time accounting snapshots are matching CTESimple 1`] = `
".--------------------------------------------------------------------------------.
|                                   CTESimple                                    |
|--------------------------------------------------------------------------------|
| # |        Label        | Actual Total Time | Total Time | Self Time | Virtual |
|---|---------------------|-------------------|------------|-----------|---------|
| 1 | Query               |                   |   102.2 ms |      0 μs |       x |
| 2 |   Planning          |                   |      51 μs |     51 μs |       x |
| 3 |   Execution         |                   |   102.2 ms |     19 μs |       x |
| 4 |     CTE Scan on foo |          102.2 ms |   102.2 ms |      7 μs |         |
| 5 |       CTE foo       |          102.2 ms |   102.2 ms |      0 μs |       x |
| 6 |         Result      |          102.2 ms |   102.2 ms |  102.2 ms |         |
'--------------------------------------------------------------------------------'"
`;

exports[`transformQueries time accounting snapshots are matching CTESleepUnion 1`] = `
".-------------------------------------------------------------------------------------------.
|                                       CTESleepUnion                                       |
|-------------------------------------------------------------------------------------------|
| #  |             Label             | Actual Total Time | Total Time | Self Time | Virtual |
|----|-------------------------------|-------------------|------------|-----------|---------|
|  1 | Query                         |                   |   303.1 ms |      0 μs |       x |
|  2 |   Planning                    |                   |     307 μs |    307 μs |       x |
|  3 |   Execution                   |                   |   302.8 ms |    152 μs |       x |
|  4 |     Limit                     |          302.7 ms |   302.7 ms |      5 μs |         |
|  5 |       CTE foo                 |          302.6 ms |   302.6 ms |      0 μs |       x |
|  6 |         Append                |          302.6 ms |   302.6 ms |      4 μs |         |
|  7 |           Result              |          101.1 ms |   101.1 ms |  101.1 ms |         |
|  8 |           Result              |          201.5 ms |   201.5 ms |  201.5 ms |         |
|  9 |           Result              |              0 μs |       0 μs |      0 μs |         |
| 10 |       InitPlan 2 (returns $1) |          101.1 ms |       7 μs |      0 μs |       x |
| 11 |         CTE Scan on foo foo_1 |          101.1 ms |       7 μs |      7 μs |         |
| 12 |       Result                  |          302.7 ms |      60 μs |     47 μs |         |
| 13 |         CTE Scan on foo       |          201.5 ms |      13 μs |     13 μs |         |
'-------------------------------------------------------------------------------------------'"
`;

exports[`transformQueries time accounting snapshots are matching PGIndexes 1`] = `
".-------------------------------------------------------------------------------------------------------------------------.
|                                                        PGIndexes                                                        |
|-------------------------------------------------------------------------------------------------------------------------|
| #  |                            Label                            | Actual Total Time | Total Time | Self Time | Virtual |
|----|-------------------------------------------------------------|-------------------|------------|-----------|---------|
|  1 | Query                                                       |                   |     9.2 ms |      0 μs |       x |
|  2 |   Planning                                                  |                   |     974 μs |    974 μs |       x |
|  3 |   Execution                                                 |                   |     8.2 ms |    135 μs |       x |
|  4 |     Nested Loop Left Join                                   |            8.1 ms |     8.1 ms |    6.9 ms |         |
|  5 |       Hash Left Join                                        |            1.2 ms |     1.2 ms |    110 μs |         |
|  6 |         Nested Loop                                         |            1.1 ms |     1.1 ms |     84 μs |         |
|  7 |           Hash Join                                         |            469 μs |     469 μs |    142 μs |         |
|  8 |             Seq Scan on pg_index x                          |             53 μs |      53 μs |     53 μs |         |
|  9 |             Hash                                            |            274 μs |     274 μs |     40 μs |         |
| 10 |               Seq Scan on pg_class c                        |            234 μs |     234 μs |    234 μs |         |
| 11 |           Index Scan using pg_class_oid_index on pg_class i |              4 μs |     520 μs |    520 μs |         |
| 12 |         Hash                                                |             16 μs |      16 μs |      6 μs |         |
| 13 |           Seq Scan on pg_namespace n                        |             10 μs |      10 μs |     10 μs |         |
| 14 |       Materialize                                           |              0 μs |       5 μs |      0 μs |         |
| 15 |         Seq Scan on pg_tablespace t                         |              5 μs |       5 μs |      5 μs |         |
'-------------------------------------------------------------------------------------------------------------------------'"
`;

exports[`transformQueries time accounting snapshots are matching RewriteTwoQueries 1`] = `
".------------------------------------------------------------------------------------------------------.
|                                          RewriteTwoQueries                                           |
|------------------------------------------------------------------------------------------------------|
| #  |                  Label                   | Actual Total Time | Total Time | Self Time | Virtual |
|----|------------------------------------------|-------------------|------------|-----------|---------|
|  1 | Queries                                  |                   |   259.0 ms |      0 μs |       x |
|  2 |   Query 1                                |                   |   130.2 ms |      0 μs |       x |
|  3 |     Planning                             |                   |      42 μs |     42 μs |       x |
|  4 |     Execution                            |                   |   130.1 ms |    1.3 ms |       x |
|  5 |       Insert on foo                      |          128.8 ms |   128.8 ms |  105.3 ms |         |
|  6 |         Function Scan on generate_series |           23.5 ms |    23.5 ms |   23.5 ms |         |
|  7 |   Query 2                                |                   |   128.8 ms |      0 μs |       x |
|  8 |     Planning                             |                   |      41 μs |     41 μs |       x |
|  9 |     Execution                            |                   |   128.8 ms |    1.4 ms |       x |
| 10 |       Insert on bar                      |          127.4 ms |   127.4 ms |  105.3 ms |         |
| 11 |         Function Scan on generate_series |           22.0 ms |    22.0 ms |   22.0 ms |         |
'------------------------------------------------------------------------------------------------------'"
`;
